Save features into PostgreSQL 12+ tables with GENERATED columns.

espinafre opened the pull request:
Fixes #32898 . GENERATED columns is a feature introduced by PostgreSQL
12, which allows column values to be generated from other columns in the
same table; this replaces the creation of triggers to, for instance,
populate a column of centroids for the stored polygons.

Quoting the PostgreSQL 12 [documentation](https://www.postgresql.org/docs/12/ddl-generated-columns.html), "A generated column cannot be written to directly. In INSERT or UPDATE commands, a value cannot be specified for a generated column, but the keyword DEFAULT may be specified."

With this commit, QGIS would use the keyword "DEFAULT". The minor drawback that I see is that users will see the string "DEFAULT" on the form field for GENERATED columns when inserting/editing a new feature, instead of the expression used to generate the table; maybe a TODO? The original expression is certainly fetched by the current code.

I've used the same table as the original reporter of the issue to test this fix; it works for me:

```sql
CREATE TABLE test_gen_col (
  id SERIAL PRIMARY KEY,
  geom GEOMETRY('Polygon', 4326) NOT NULL,
  cent GEOMETRY('Point') GENERATED ALWAYS AS ( st_centroid(geom) ) STORED
);
```

Should this be backported to 3.12 and 3.10?

espinafre committed:
Save features into tables with GENERATED columns.

Fixes #32898 . GENERATED columns is a feature introduced by PostgreSQL
12, which allows column values to be generated from other columns in the
same table; this replaces the creation of triggers to, for instance,
populate a column of centroids for the stored polygons.

espinafre commented:
The tests run correctly on my local system (PostgreSQL 12.0); since Travis seems to use PostgreSQL 11, I'm setting up a new environment with that version and test to figure what went wrong.

espinafre committed:
Handling PostgreSQL <= 11.

espinafre commented:
@jef-n , since you worked on handling identities (#9735), can you take a look at this, as it uses some of your ideas? Also, for some reason the github interface doesn't allow me to request a review; would you please step in? Thank you very much.

espinafre commented:
Is it possible to manually trigger the Azure build again?

espinafre committed:
Mechanism to test features on PostgreSQL 12+

espinafre commented:
WIP because the local test with PostgreSQL 12 fails (this won't be caught by current Travis config, which still uses PostgreSQL 11).

espinafre committed:
Handling of GENERATED columns for all column types

This enables the use of GENERATED columns, a feature introduced by
PostgreSQL 12, with QGIS. This features replaces triggers for common
operations, such as calculating centroids and areas of polygons.

espinafre committed:
Tests for PostgreSQL 12+ GENERATED columns

These tests exercise the use of GENERATED columns, a feature which was
introduced by PostgreSQL 12+. They won't be run if the PostgreSQL server
against which we are testing is older than version 12.

espinafre committed:
Correct handling of generated values

espinafre commented:
All green here. @nyalldawson , @m-kuhn , @jef-n , could you please take a look? Is there a way for the CI systems to check the changes I've made on the testing scripts to support PostgreSQL 12+-only features? Also, any chance to get this on next 3.12 release? Thank you all very much.

espinafre commented:
Hi, @nyalldawson , what do you think of this?

nyalldawson commented:
I'm not the best person to review this sorry - maybe @m-kuhn or @elpaso could take a look?

m-kuhn commented:
Thanks for looking into this. GENERATED columns are a really nice concept.

I wonder if this already handles updates? I see it only takes care of the `addFeatures()` method.
In general, could these values just be skipped everywhere (rather than using the `DEFAULT` keyword)? This would (slightly) reduce network I/O.

This way, the default value would be completely free to be used for something else than a static `DEFAULT` value and could instead be used to communicate to the user what calculation will be performed.

Ideally the field would also be set to read only for the attribute forms, but I don't think there's an API for that already.

espinafre commented:
Hi, @m-kuhn , it does handle updates, the trick is on the defaultValueClause() method, we use the DEFAULT keyword in all cases. 
The PostgreSQL docs says we must either use DEFAULT or omit the column, which I didn't find a way to do.

m-kuhn commented:
Let's say you have

```sql
CREATE TABLE test_gen_col (
  id SERIAL PRIMARY KEY,
  geom GEOMETRY('Polygon', 4326) NOT NULL,
  x integer GENERATED ALWAYS AS ( st_x(geom) ) STORED
);
```

```
feature = next(layer.getFeatures())
feature.setGeometry(QgsGeometry.fromWkt('Point(1 2)'))
feature['x'] = 5
layer.updateFeature(feature)
```

Is `DEFAULT` used? I might be missing something, but I thought the defaultValueClause() is mostly used by the attribute form to initialize new features.

espinafre commented:
Yes, DEFAULT is not used on UPDATE, of course. The PostgreSQL definition says, however, that the column must not be manually written in any way, either by ommiting it or by using the DEFAULT keyword. For INSERTs (addFeatures()), the defaultValueClause() generates the keyword DEFAULT; for UPDATEs (changeAttributeValues()), nothing is changed.

 Changing your test table definition a little (st_x/st_y applies to point):

```sql
CREATE TABLE test_gen_col3 (
  id SERIAL PRIMARY KEY,
  geom GEOMETRY('Point', 4326) NOT NULL,
  x integer GENERATED ALWAYS AS ( st_x(geom) ) STORED
);
```

Inserting a point generates the following:

```sql
INSERT INTO "public"."test_gen_col3" ("geom
","x") VALUES (st_geomfromwkb($1::bytea,4326),DEFAULT) RETURNING "id"
```

Trying to change the X value yields the following error:

```
src/core/qgsmessagelog.cpp:29 : (logMessage) [2ms] 2020-05-23T17:13:45 Messages[2] Layer test_gen_col3 : PostGIS error while changing attributes: ERROR:  column "x" can only be updated to DEFAULT
DETAIL:  Column "x" is a generated column.
```
This is the query that gets sent to the DBMS, which generates the above error:

```sql
UPDATE "public"."test_gen_col3" SET "x"=31 WHERE "id"=1
```
This is also written to the user console as a critical error.

The correct solution, I think, is to make generated fields (and other fields with DEFAULT values, perhaps) read-only, but I'd have to study the code more deeply in order to better understand how we can send this information from the provider to the UI.

m-kuhn commented:
I think there would be a preferable initial implementation.

IMHO, the first thing to do to implement generated columns is that the postgres provider will need to check with every insert or update if any of the involved attributes is generated and replace (or omit them). This can be done without touching the defaultValueClause(), directly where the insert or update statement is created. This should be a bit more robust.

Making the attributes read only would be an additional nice move as you say, but is not required for a minimal implementation.

Another nice move would be to consider that the geometry might also be generated (e.g. from x and y attributes), but that's also nice to have.

espinafre commented:
> Another nice move would be to consider that the geometry might also be generated (e.g. from x and y attributes), but that's also "nice to have".

I was going to reply that generated geometry is what motivated this (centroid from polygon) and is what I've put on the test cases, but it has just dawned on me that you mean that the main geometry of the table might be generated. I think it can be done with the current approach, but the user interface needs some polishing; he must have a way to insert a feature without a geometry.

> IMHO, the first thing to do to implement generated columns is that the postgres provider will need to check with every insert or update if any of the involved attributes is generated and replace (or omit them). 

Give me a couple of days and I will try this approach. Also, I've just realized that on my test cases there is nothing that covers UPDATEs, I'll do that as well.

m-kuhn commented:
> but the user interface needs some polishing; he must have a way to insert a feature without a geometry.

There's already an "add row" button in the attribute table for this purpose.

espinafre committed:
Changed insert/update of GENERATED fields

Changed the way GENERATED fields are inserted/updated. Previously we
used the DEFAULT keyword for INSERTing; now, we expose the generating
expression to the user, but ommit the field when inserting or updating a
feature.

espinafre commented:
Okay, here ( daa3bba ) are the changes according to @m-kuhn's suggestion, that is, checking whether each field is generated, and omit those that are in case of INSERT or UPDATEs. defaultValueClause() now returns the expression used to generate the field, instead of the keyword "DEFAULT"; this way, the user can see what is happening.

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -2313,6 +2338,13 @@ bool QgsPostgresProvider::addFeatures( QgsFeatureList &flist, Flags flags )
 
       QString fieldname = mAttributeFields.at( idx ).name();
+
+      if ( mGeneratedValues.contains( idx ) && !mGeneratedValues.value( idx, QString() ).isEmpty() )
```
```suggestion
      if ( !mGeneratedValues.value( idx ).isEmpty() )
```

That should be sufficient

espinafre replied:
Sure, resolved with commit 793409c.

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -2339,7 +2371,12 @@ bool QgsPostgresProvider::addFeatures( QgsFeatureList &flist, Flags flags )
       {
-        if ( qgsVariantEqual( v, defVal ) )
+        // generated values override everything.
+        if ( !mGeneratedValues[idx].isEmpty() )
```
```suggestion
        if ( !mGeneratedValues.value( idx ).isEmpty() )
```

espinafre replied:
Resolved with 793409c (condition was deleted).

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -2339,7 +2371,12 @@ bool QgsPostgresProvider::addFeatures( QgsFeatureList &flist, Flags flags )
+        // generated values override everything.
+        if ( !mGeneratedValues[idx].isEmpty() )
+        {
+          values += delim + defVal;
```
```suggestion
          values += delim + QStringLiteral( "DEFAULT" );
```

We need to specify `DEFAULT` here, not the calculation expression, right?

espinafre replied:
Actually this is dead code, since we don't use DEFAULT anymore, and skip the GENERATED field entirely (line 2342)

espinafre committed:
Code review.

espinafre committed:
Don't assume pg_attribute.attgenerated will always have "s"

The value for pg_attribute.attgenerated will hold "s" for stored
generated fields, but this might change in future releases of
PostgreSQL. So, we just check for the presence of a value in that field.

espinafre committed:
Added test cases for updating of tables with GENERATED fields.

espinafre committed:
Make flake8 happy (formatting).

espinafre commented:
Hi, 6374338 adds the tests for changing of attributes and geometries for tables with GENERATED columns.
@m-kuhn , is pushError() on src/providers/postgres/qgspostgresprovider.cpp#L2955-L2959 
 
 the right way to warn the user that he is trying to do something wrong?

espinafre commented:
Hi, @m-kuhn , are you okay with these changes? Can we merge?

I'd like to call your attention to the fact that this new code path is not yet tested on Travis, because it still uses PostgreSQL 11.

I've put a PostgreSQL version check on the loading script, and also on the test method itself, so that if it detects we are running against PostgreSQL older than 12, the method immediately returns. That's because the new syntax is incompatible with older PostgreSQL versions. However, as I've stated earlier, the tests run correctly on my local system, which has PostgreSQL 12.3.

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -881,15 +881,17 @@ bool QgsPostgresProvider::loadFields()
-              " WHERE attrelid IN %2"
-            ).arg( connectionRO()->pgVersion() >= 100000 ? QStringLiteral( ", attidentity" ) : QString() ).arg( tableoidsFilter );
+              " WHERE attrelid IN %3"
+            ).arg( connectionRO()->pgVersion() >= 100000 ? QStringLiteral( ", attidentity" ) : QString() )
```
```suggestion
            ).arg( connectionRO()->pgVersion() >= 100000 ? QStringLiteral( ", attidentity" ) : QString(),
```

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -881,15 +881,17 @@ bool QgsPostgresProvider::loadFields()
-            ).arg( connectionRO()->pgVersion() >= 100000 ? QStringLiteral( ", attidentity" ) : QString() ).arg( tableoidsFilter );
+              " WHERE attrelid IN %3"
+            ).arg( connectionRO()->pgVersion() >= 100000 ? QStringLiteral( ", attidentity" ) : QString() )
+            .arg( connectionRO()->pgVersion() >= 120000 ? QStringLiteral( ", attgenerated" ) : QString() )
```
```suggestion
            connectionRO()->pgVersion() >= 120000 ? QStringLiteral( ", attgenerated" ) : QString(),
```

espinafre replied:
Addressed by 5a56586

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -881,15 +881,17 @@ bool QgsPostgresProvider::loadFields()
+              " WHERE attrelid IN %3"
+            ).arg( connectionRO()->pgVersion() >= 100000 ? QStringLiteral( ", attidentity" ) : QString() )
+            .arg( connectionRO()->pgVersion() >= 120000 ? QStringLiteral( ", attgenerated" ) : QString() )
+            .arg( tableoidsFilter );
```
```suggestion
            tableoidsFilter );
```

A single `.arg` is more performant.

espinafre replied:
Addressed by 5a56586

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -2313,6 +2338,13 @@ bool QgsPostgresProvider::addFeatures( QgsFeatureList &flist, Flags flags )
+
+      if ( !mGeneratedValues.value( idx, QString() ).isEmpty() )
+      {
+        QgsDebugMsg( QStringLiteral( "Skipping field %1 (idx %2) which is GENERATED." ).arg( fieldname, idx ) );
```
```suggestion
        QgsDebugMsg( QStringLiteral( "Skipping field %1 (idx %2) which is GENERATED." ).arg( fieldname, QString::number( idx ) ) );
```

This was calling the wrong overload. https://doc.qt.io/qt-5/qstring.html#arg

espinafre replied:
Addressed by 5a56586

m-kuhn reviewed (commented):
Looks good to me now, thanks. Just some minor string formatting left.

For the tests, we could consider to upgrade postgres on travis to 12, I'm not sure what other side effects that will have.

Concerning error reporting. Is making this message as prominent as putting it on the message bar justified? Or would it be better in the message log?
Basically, is it something which is the result of a user mistake and he needs to be informed (message bar) or more a code problem / unimportant side effect (message log).
I'm fine with both.

espinafre committed:
Code review && QgsLogger::warning if user tries to insert/change generated field.

m-kuhn commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -2341,7 +2341,7 @@ bool QgsPostgresProvider::addFeatures( QgsFeatureList &flist, Flags flags )
       if ( !mGeneratedValues.value( idx, QString() ).isEmpty() )
       {
-        QgsDebugMsg( QStringLiteral( "Skipping field %1 (idx %2) which is GENERATED." ).arg( fieldname, idx ) );
+        QgsLogger::warning( tr( "Skipping field %1 which is GENERATED." ).arg( fieldname ) );
```
This will lead to warnings on every "add feature" with generated columns, correct?

espinafre replied:
Yes, correct.

espinafre replied:
I've removed the "idx" from the message because that shouldn't matter to the user, and developers can easily find out the field index if needed.

m-kuhn replied:
I would completely remove this (or make it a debug message as before), as it currently is, the log will be spammed and there is no way out.

espinafre replied:
Of course, this hadn't ocurred to me; I was thinking about inserting features only with the user-editable fields. Reverted to debug message in 0d1a4c5.

espinafre commented:
Commit 5a56586 addresses issues pointed out by @m-kuhn , and also replaces pushError() with QgsLogger::warning. I've done this because the user has no way, in principle, to detect that a field is auto-generated by the database, so I deem it is not a user error if he tries to change a generated field. However, the user has to know that a field was not updated, and we warn him so.

Below are the last few lines of the output of `ctest -V -R PyQgsPostgresProvider` locally against PostgreSQL 12.3 (I have the full logs if needed):

```
591: PostGIS(1): FAILURE: Field -1 not found.
591: PostGIS(1): FAILURE: Field 1000 not found.
591: PostGIS(1): FAILURE: Field -1 not found..src/core/qgsmessagelog.cpp:29 : (logMessage) [2ms] 2020-06-02T10:00:20 PostGIS[1] FAILURE: Field -1 not found.
591: src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] 2020-06-02T10:00:20 PostGIS[1] FAILURE: Field 100001 not found.
591: ...
591: ----------------------------------------------------------------------
591: Ran 233 tests in 57.896s
591: 
591: OK
591: 
591: PostGIS(1): FAILURE: Field 1000 not found.
591: PostGIS(1): FAILURE: Field -1 not found.
591: PostGIS(1): FAILURE: Field 100001 not found.
1/1 Test #591: PyQgsPostgresProvider ............   Passed   69.19 sec

The following tests passed:
        PyQgsPostgresProvider

100% tests passed, 0 tests failed out of 1

Total Test time (real) =  69.64 sec
```

espinafre committed:
Avoid spamming warnings on INSERT with generated fields.

Thanks to M. Kuhn for pointing out that the old code would drown the
console with warnings if one tries to load data into a table with
generated fields.

m-kuhn reviewed (approved):
Everything has been addressed and code is looking good.
Let's give headway to PG12 goodness :+1: 

Thanks a lot @espinafre for the pleasurable work

espinafre commented:
It is I who have to thank you @m-kuhn for your patience and guidance.

Speaking of which, is there something already in place so we can start working on marking read-only fields as such to the user?

Do you think it would be good to backport to LTR? Or label this as a feature so my name can get included in the QGIS hall of fame? :smile: 

Almost everything green here, except for the 32 bit Azure build (which is failing for everyone else too).

espinafre committed:
Fixed missing cast to string in debug message.

This slipped through on last commit.

m-kuhn commented:
> Speaking of which, is there something already in place so we can start working on marking read-only fields as such to the user?

Not that I was aware of it, but possibly ...

> Or label this as a feature so my name can get included in the QGIS hall of fame?

If you can paste the text for the changelog including the name that should appear and ideally a screenshot, I'll be happy to put it to [the visual changelog](https://changelog.qgis.org/en/qgis/version/3.14/) for you.

Regarding backport, it would be nice to support this, but I'd give it at least one month to cool down and see if it has any unexpected side effects.

espinafre commented:
@m-kuhn , for the visual changelog, in the "Data Providers" section, right after Pasotti's "PG raster expose set filter to app", something like:

> ### PG: save features into tables with generated fields
> 
> Allow updating/inserting features in PostgreSQL tables with generated fields. GENERATED columns is a feature introduced by PostgreSQL 12, which allows column values to be generated from other columns in the same table; this replaces the creation of triggers to, for instance,
> automatically populate and keep up-to-date a column of centroids or areas for the polygons in the main geometry column.
> By [José de Paula Rodrigues N. Assis](https://api.github.com/users/espinafre)

Unfortunately, I couldn't come up with good screenshots; all I could think about is something showing the error before the change, and filled fields after it, but I don't think these are good enough material.

Might my name be added to the Contributors list in the About dialog? Thank you very much.

m-kuhn commented:
Here you go:

https://changelog.qgis.org/en/qgis/version/3.14/#postgres-save-features-into-tables-with-generated-

Help yourself into the about dialog by creating a pull request for resources/data/contributors.json
